sdks/java/core/src/main/java/org/apache/beam/sdk/util/ShardedKey.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.StructuredCoder;
+import org.apache.beam.sdk.util.common.ElementByteSizeObserver;
+
+/** A sharded key consisting of a user key and an opaque shard id represented by bytes. */
+public class ShardedKey<K> {
+  private static final byte[] EMPTY_SHARD_ID = new byte[0];
+
+  private final K key;
+  private final byte[] shardId;
+
+  protected ShardedKey(K key, byte[] shardId) {
+    this.key = key;
+    this.shardId = shardId;
+  }
+
+  /** Creates a ShardedKey with given key and shard id. Shard id must not be null. */
+  public static <K> ShardedKey<K> of(K key, byte[] shardId) {
+    checkArgument(key != null, "Key should not be null!");
+    checkArgument(shardId != null, "Shard id should not be null!");
+    return new ShardedKey<K>(key, shardId);
+  }
+
+  public K getKey() {
+    return key;
+  }
+
+  @Override
+  public String toString() {
+    return "ShardedKey{" + "key=" + key + ", shardId=" + Arrays.toString(shardId) + "}";
+  }
+
+  @Override
+  public boolean equals(Object o) {
```

Comment:
FYI, this equals method assumes the key type has a valid equals method and generally shouldn't be used. We should use structuralValue anywhere for comparisons.

Reply:
I don't want you to change it, its just an FYI.

Reply:
Ack.